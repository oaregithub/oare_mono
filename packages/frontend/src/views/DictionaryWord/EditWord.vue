<template>
  <div>
    <OareLoaderButton
      color="primary"
      class="mb-5 test-save-btn"
      :loading="saveLoading"
      @click="saveEdits"
      >Save Edits</OareLoaderButton
    >
    <OareLabel>Word</OareLabel>
    <v-col cols="11" sm="7" lg="5">
      <v-text-field
        v-model="localWordInfo.word"
        outlined
        class="mb-4 test-edit-word"
      />
    </v-col>
    <OareLabel>Translations</OareLabel>

    <div
      v-for="(translation, idx) in localWordInfo.translations"
      :key="idx"
      class="d-flex align-start"
    >
      <span class="mt-6">{{ idx + 1 }}</span>
      <v-col cols="11" sm="7" lg="5" class="mb-n4">
        <v-text-field
          v-model="localWordInfo.translations[idx].translation"
          outlined
        />
      </v-col>
      <div class="d-flex mt-5">
        <v-btn v-if="idx !== 0" icon @click="moveTranslation(idx, idx - 1)">
          <v-icon>mdi-chevron-up</v-icon>
        </v-btn>
        <v-btn
          v-if="idx !== localWordInfo.translations.length - 1"
          icon
          @click="moveTranslation(idx, idx + 1)"
        >
          <v-icon>mdi-chevron-down</v-icon>
        </v-btn>
        <v-btn icon @click="removeTranslation(idx)">
          <v-icon>mdi-trash-can-outline</v-icon>
        </v-btn>
      </div>
    </div>
    <v-btn color="primary" @click="addTranslation" text>
      <v-icon>mdi-plus</v-icon>
      Add translation
    </v-btn>

    <v-snackbar v-model="undoTranslateSnackbar">
      Translation deleted
      <template #action="{ attrs }">
        <v-btn text color="info" v-bind="attrs" @click="undoDeleteTranslation">
          Undo
        </v-btn>
      </template>
    </v-snackbar>
  </div>
</template>
<script lang="ts">
import { defineComponent, PropType, ref, Ref } from '@vue/composition-api';
import { WordWithForms } from '@/types/dictionary';
import _ from 'lodash';
import defaultServerProxy from '@/serverProxy';
import defaultActions from '@/globalActions';

export interface DeletedTranslation {
  index: number;
  translation: {
    uuid: string;
    translation: string;
  };
}

export default defineComponent({
  name: 'EditWord',
  props: {
    uuid: {
      type: String,
      required: true,
    },
    wordInfo: {
      type: Object as PropType<WordWithForms>,
      required: true,
    },
    serverProxy: {
      type: Object as PropType<typeof defaultServerProxy>,
      default: () => defaultServerProxy,
    },
    actions: {
      type: Object as PropType<typeof defaultActions>,
      default: () => defaultActions,
    },
  },
  setup({ uuid, wordInfo, serverProxy, actions }, { emit }) {
    const loading = ref(true);
    const saveLoading = ref(false);
    const localWordInfo: Ref<WordWithForms> = ref(_.cloneDeep(wordInfo));
    const undoTranslateSnackbar = ref(false);
    const deletedTranslation: Ref<DeletedTranslation | null> = ref(null);

    const removeTranslation = (index: number) => {
      deletedTranslation.value = {
        index,
        translation: localWordInfo.value.translations[index],
      };
      localWordInfo.value.translations.splice(index, 1);
      undoTranslateSnackbar.value = true;
    };

    const moveTranslation = (oldIndex: number, newIndex: number) => {
      const translations = [...localWordInfo.value.translations];

      const temp = translations[oldIndex];
      translations[oldIndex] = translations[newIndex];
      translations[newIndex] = temp;

      localWordInfo.value.translations = translations;
    };

    const addTranslation = () => {
      localWordInfo.value.translations.push({
        uuid: '',
        translation: '',
      });
    };

    const undoDeleteTranslation = () => {
      if (deletedTranslation.value) {
        const { index, translation } = deletedTranslation.value;

        localWordInfo.value.translations.splice(index, 0, translation);
      }
      undoTranslateSnackbar.value = false;
    };

    const saveEdits = async () => {
      saveLoading.value = true;
      const payload = {
        word: localWordInfo.value.word,
        translations: localWordInfo.value.translations,
      };

      try {
        const { translations } = await serverProxy.editWord(uuid, payload);
        localWordInfo.value.translations = translations;
        emit('update:wordInfo', {
          ...localWordInfo.value,
          ...payload,
        });
        actions.showSnackbar('Edit saved');
      } catch {
        actions.showErrorSnackbar('Error saving word');
      } finally {
        saveLoading.value = false;
      }
    };

    return {
      loading,
      localWordInfo,
      removeTranslation,
      moveTranslation,
      addTranslation,
      undoTranslateSnackbar,
      undoDeleteTranslation,
      saveEdits,
      saveLoading,
    };
  },
});
</script>
