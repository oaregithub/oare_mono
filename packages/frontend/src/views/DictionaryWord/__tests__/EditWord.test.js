import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import { mount, createLocalVue } from '@vue/test-utils';
import EditWord from '../EditWord.vue';
import flushPromises from 'flush-promises';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('EditWord test', () => {
  const mockServer = {
    editWord: jest.fn().mockResolvedValue({
      translations: [],
    }),
  };
  const mockSnackbar = {
    showSnackbar: jest.fn(),
    showErrorSnackbar: jest.fn(),
  };
  const mockProps = {
    uuid: 'uuid',
    wordInfo: {
      word: 'test word',
      forms: [],
      partsOfSpeech: [],
      verbalThematicVowelTypes: [],
      specialClassifications: [],
      translations: [],
    },
    serverProxy: mockServer,
    actions: mockSnackbar,
  };
  const createWrapper = (props = mockProps) =>
    mount(EditWord, {
      vuetify,
      localVue,
      propsData: {
        ...mockProps,
        ...props,
      },
    });

  it('matches snapshot', () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it('edits word spelling', async () => {
    const wrapper = createWrapper();
    await wrapper.find('.test-edit-word input').setValue('new word');
    await wrapper.find('.test-save-btn').trigger('click');
    await flushPromises();
    expect(mockServer.editWord).toHaveBeenCalledWith('uuid', {
      word: 'new word',
      translations: [],
    });
    expect(mockSnackbar.showSnackbar).toHaveBeenCalledWith('Edit saved');
  });

  it("shows error when edit doesn't work", async () => {
    const errorServer = {
      editWord: jest.fn().mockRejectedValue(null),
    };
    const wrapper = createWrapper({
      serverProxy: errorServer,
    });

    await wrapper.find('.test-edit-word input').setValue('new word');
    await wrapper.find('.test-save-btn').trigger('click');
    await flushPromises();
    expect(mockSnackbar.showErrorSnackbar).toHaveBeenCalledWith(
      'Error saving word'
    );
  });
});
