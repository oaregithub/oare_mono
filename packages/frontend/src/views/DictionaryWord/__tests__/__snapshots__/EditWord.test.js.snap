// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditWord test matches snapshot 1`] = `
<div><button type="button" class="mb-5 test-save-btn v-btn v-btn--contained theme--light v-size--default primary" color="primary"><span class="v-btn__content">Save Edits</span></button>
  <div class="font-weight-bold mb-2">Word</div>
  <div class="col-sm-7 col-lg-5 col-11">
    <div class="v-input mb-4 test-edit-word v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend style="width: 0px;"><span>â€‹</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input id="input-6" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="font-weight-bold mb-2">Translations</div> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-plus theme--light"></i>
    Add translation
  </span></button>
  <div class="v-snack v-snack--bottom v-snack--has-background" style="padding-bottom: 0px; padding-top: 0px;">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet theme--dark" style="display: none;">
        <div role="status" aria-live="polite" class="v-snack__content">
          Translation deleted
        </div>
        <div class="v-snack__action "><button type="button" class="v-btn v-btn--flat v-btn--text theme--dark v-size--default info--text v-snack__btn"><span class="v-btn__content">
        Undo
      </span></button></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
